src/OpenTelemetry/Metrics/Tags.cs

```diff
@@ -52,33 +34,102 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
 
     public readonly bool Equals(Tags other)
     {
-        var length = this.KeyValuePairs.Length;
+        var ourKvps = this.KeyValuePairs;
+        var theirKvps = other.KeyValuePairs;
 
-        if (length != other.KeyValuePairs.Length)
+        if (ourKvps.Length != theirKvps.Length)
         {
             return false;
         }
 
-        for (int i = 0; i < length; i++)
+#if NET6_0_OR_GREATER
+        // Note: This loop uses unsafe code (pointers) to elide bounds checks on
+        // two arrays we know to be of equal length.
+        var cursor = ourKvps.Length;
+        if (cursor > 0)
+        {
+            ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
```

Comment:
Is this and the `Unsafe.Add` calls vulnerable to garbage collector relocating this array when compacting memory?

Reply:
AFAIK they are managed pointers meaning the GC is aware of them.

I ran this pattern by @stephentoub a while back. IIRC the answer was it is safe to do so long as you know for sure the array references won't change out from under you and they are indeed the same length.

Reply:
If done correctly, it's valid. These are managed refs, so if the GC were to move things around, it would also update the values stored in the refs to keep them valid. The concern with code like this is it's easy to accidentally get it wrong, at which point you're subject to all the same kinds of memory safety problems as in languages like C++. As a simple for loop, the JIT would already be eliminating the bounds check on accessing the elements from whichever span was used in the for loop condition, so at best this is only eliminating the other half of the bounds checks. You'll want to be really, really sure it's worth it.